### PR TITLE
Virtual wallet

### PR DESCRIPTION
**Pull request by verdefred**
With this PR we want to add:

a new contract called VirtualTAL where we'll have:
* Mapping from an address to a value in virtual TAL
* Function for admin to add TAL to that mapping
* Function for admin to remove TAL to that mapping
* Function to view the value of virtual TAL for any address

a new contract called StakingV2 that inherits the Staking contract where we'll have:
* Function to create a stake with virtual TAL
* Function to sell talent tokens for virtual TAL

### CODE REVIEW

File: contracts/StakingV2.sol [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    /// @param _amount The talent tokens amount
+    /// @return bool
+    function sellTalentTokenWithVirtualTAL(address _talentTokenAddress, uint256 _amount)
+        public
+        onlyWhileStakingEnabled
+        stablePhaseOnly
+        returns (bool)
+    {
+        require(!disabled, "1");
+        require(IERC20(_talentTokenAddress).balanceOf(msg.sender) >= _amount, "3");
```
Comment (RubenSousaDinis) [inline contracts/StakingV2.sol:53]: why 3?

File: contracts/StakingV2.sol [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    /// @param _talentTokenAddress The talent address
+    /// @param _amount The talent tokens amount
+    /// @return bool
+    function sellTalentTokenWithVirtualTAL(address _talentTokenAddress, uint256 _amount)
+        public
+        onlyWhileStakingEnabled
+        stablePhaseOnly
+        returns (bool)
+    {
+        require(!disabled, "1");
```
Comment (RubenSousaDinis) [inline contracts/StakingV2.sol:52]: why 1?

File: contracts/staking_v2_helpers/VirtualTALHelper.sol [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    function setVirtualTALAddress() external view returns (uint256);
+}
+
+/// @title A two-phase contract, starting with a stable coin and proceeding to
+//    a alternative token once possible.
+///
+/// @notice Since the first phase of staking will be done in a USD-pegged
+///   stable-coin, we need a mechanism to later /   switch to the TAL token, while
+///   also converting any initial USD stakes to TAL, given a pre-determined rate
+abstract contract VirtualTALHelper is Initializable, AccessControlEnumerableUpgradeable {
```
Comment (RubenSousaDinis) [inline contracts/staking_v2_helpers/VirtualTALHelper.sol:25]: Missing interface? IVirtualTALHelper

File: contracts/StakingV2.sol [hunk 4]
```diff
@@ -0,0 +1,61 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.7;
```
Comment (RubenSousaDinis) [inline contracts/StakingV2.sol:2]: Should we think about upgrading?

File: contracts/Staking.sol [hunk 5]
```diff
@@ -567,7 +576,9 @@ contract Staking is
         uint256 proportion = (_talentAmount * MUL) / stake.talentAmount;
```
Comment (RubenSousaDinis) [inline contracts/Staking.sol:576]: start using math safe?

**Review by RubenSousaDinis — COMMENTED** [review]
Added some initial comments

File: contracts/StakingV2.sol (hunk 1 above)
Comment (verdefred) [reply contracts/StakingV2.sol:53]: just an intermediate step to not have a smart contract too big. will change in the next commit 🙏

File: contracts/StakingV2.sol (hunk 2 above)
Comment (verdefred) [reply contracts/StakingV2.sol:52]: just an intermediate step to not have a smart contract too big. will change in the next commit 🙏

File: contracts/staking_v2_helpers/VirtualTALHelper.sol (hunk 3 above)
Comment (verdefred) [reply contracts/staking_v2_helpers/VirtualTALHelper.sol:25]: it's in the lines above

File: contracts/StakingV2.sol (hunk 4 above)
Comment (verdefred) [reply contracts/StakingV2.sol:2]: This comes with some issues for the rest of the contracts but we can see if it makes sense to do it!

File: contracts/Staking.sol (hunk 5 above)
Comment (verdefred) [reply contracts/Staking.sol:576]: yeah, absolutely! thanks for the reminder

File: contracts/season3/StakingV3.sol [hunk 6]
```diff
@@ -0,0 +1,971 @@
...
+    }
+
+    /// Redeems rewards since last checkpoint and withdraws to Virtual TAL
+    ///
+    /// @return true if operation succeeds
+    function claimRewardsToVirtualTAL() public stablePhaseOnly returns (bool) {
+        claimRewardsOnBehalf(msg.sender);
+
+        return true;
+    }
```
Comment (RubenSousaDinis) [inline contracts/season3/StakingV3.sol:325]: We can save some lines here:

```suggestion
    function claimRewardsToVirtualTAL() public stablePhaseOnly returns (bool) {
        return claimRewardsOnBehalf(msg.sender);
    }
```

File: contracts/season3/StakingV3.sol [hunk 7]
```diff
@@ -0,0 +1,971 @@
...
+        require(stake.talentAmount >= _talentAmount);
+
+        // calculate TAL amount proportional to how many talent tokens are
+        // being deposited if stake has 100 deposited TAL + 1 TAL earned from
+        // rewards, then returning 1 Talent Token should result in 50.5 TAL
+        // being returned, instead of the 50 that would be given under the set
+        // exchange rate
+        uint256 proportion = (_talentAmount * IRewardCalculatorV2(rewardCalculator).mul()) / stake.talentAmount;
+        uint256 tokenAmount = (stake.tokenAmount * proportion) / IRewardCalculatorV2(rewardCalculator).mul();
+
```
Comment (RubenSousaDinis) [inline contracts/season3/StakingV3.sol:591]: SafeMath?

File: contracts/season3/StakingV3.sol [hunk 8]
```diff
@@ -0,0 +1,971 @@
...
+            totalTalentTALInvested
+        );
+
+        rewardsGiven += stakerRewards + talentRewards;
+        globalStake.S = S;
+        globalStake.lastCheckpointAt = block.timestamp;
+        if (totalTALInvested == 0) {
+            talentS = 0;
+        } else {
+            talentS = (talentS + (talentRewards * IRewardCalculatorV2(rewardCalculator).mul()) / totalTALInvested);
```
Comment (RubenSousaDinis) [inline contracts/season3/StakingV3.sol:760]: SafeMath?

File: contracts/season3/StakingV3.sol [hunk 9]
```diff
@@ -0,0 +1,971 @@
...
+    function calculateEstimatedReturns(address _owner, uint256 _currentTime)
+        public
+        view
+        returns (uint256 stakerRewards, uint256 talentRewards)
+    {
+        GlobalStakeData storage globalStake = globalStakes[_owner];
+        uint256 newS = S +
+            (IRewardCalculatorV2(rewardCalculator).calculateGlobalReward(start, end, SAt, _currentTime, rewardsMax)) /
+            totalAdjustedShares;
+
```
Comment (RubenSousaDinis) [inline contracts/season3/StakingV3.sol:799]: SafeMath?

**Review by RubenSousaDinis — COMMENTED** [review]
Some small comments

**Pull request merged**